File: backend/scripts/lint.py
#!/usr/bin/env python3
"""
Comprehensive linting script for Gene Curator backend.

This script runs all linting tools in the correct order and provides
a unified interface for code quality checks.
"""

import subprocess
import sys
from pathlib import Path
from typing import List, Optional


def run_command(cmd: List[str], description: str) -> bool:
    """Run a command and return True if successful."""
    print(f"🔍 {description}...")
    try:
        result = subprocess.run(
            cmd,
            cwd=Path(__file__).parent.parent,
            capture_output=True,
            text=True,
            check=False,
        )

        if result.returncode == 0:
            print(f"✅ {description} passed")
            if result.stdout.strip():
                print(f"   Output: {result.stdout.strip()}")
            return True
        else:
            print(f"❌ {description} failed")
            if result.stdout.strip():
                print(f"   stdout: {result.stdout.strip()}")
            if result.stderr.strip():
                print(f"   stderr: {result.stderr.strip()}")
            return False

    except FileNotFoundError:
        print(f"❌ {description} failed - command not found: {' '.join(cmd)}")
        return False


def main() -> int:
    """Run all linting checks."""
    print("🚀 Starting Gene Curator Backend Linting Suite")
    print("=" * 60)

    checks = [
        (["uv", "tree"], "uv dependency tree check"),
        (["uv", "run", "ruff", "check", "app/"], "Ruff linting"),
        (["uv", "run", "ruff", "format", "--check", "app/"], "Ruff formatting check"),
        (["uv", "run", "mypy", "app/"], "MyPy type checking"),
        (["uv", "run", "bandit", "-r", "app/", "-f", "json"], "Bandit security check"),
    ]

    failed_checks = []

    for cmd, description in checks:
        if not run_command(cmd, description):
            failed_checks.append(description)

    print("\n" + "=" * 60)

    if failed_checks:
        print(f"❌ {len(failed_checks)} check(s) failed:")
        for check in failed_checks:
            print(f"   - {check}")
        print("\n💡 Run individual tools to see detailed error messages:")
        print("   - uv tree")
        print("   - uv run ruff check app/ tests/")
        print("   - uv run ruff format --diff app/ tests/")
        print("   - uv run mypy app/")
        print("   - uv run bandit -r app/")
        return 1
    else:
        print("✅ All linting checks passed!")
        return 0


if __name__ == "__main__":
    sys.exit(main())

File: frontend/scripts/lint.js
#!/usr/bin/env node
/**
 * Comprehensive linting script for Gene Curator frontend.
 *
 * This script runs all linting tools in the correct order and provides
 * a unified interface for code quality checks.
 */

import { execSync } from 'child_process'
import { fileURLToPath } from 'url'
import { dirname, join } from 'path'

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)
const projectRoot = join(__dirname, '..')

/**
 * Run a command and return success status
 */
function runCommand(cmd, description) {
  console.log(`🔍 ${description}...`)
  try {
    const result = execSync(cmd, {
      cwd: projectRoot,
      stdio: 'pipe',
      encoding: 'utf-8'
    })

    console.log(`✅ ${description} passed`)
    if (result && result.trim()) {
      console.log(`   Output: ${result.trim()}`)
    }
    return true
  } catch (error) {
    console.log(`❌ ${description} failed`)
    if (error.stdout && error.stdout.trim()) {
      console.log(`   stdout: ${error.stdout.trim()}`)
    }
    if (error.stderr && error.stderr.trim()) {
      console.log(`   stderr: ${error.stderr.trim()}`)
    }
    return false
  }
}

function main() {
  console.log('🚀 Starting Gene Curator Frontend Linting Suite')
  console.log('='.repeat(60))

  const checks = [
    ['npm run lint:check', 'ESLint linting'],
    ['npm run format:check', 'Prettier formatting check']
  ]

  const failedChecks = []

  for (const [cmd, description] of checks) {
    if (!runCommand(cmd, description)) {
      failedChecks.push(description)
    }
  }

  console.log(`\n${'='.repeat(60)}`)

  if (failedChecks.length > 0) {
    console.log(`❌ ${failedChecks.length} check(s) failed:`)
    failedChecks.forEach(check => {
      console.log(`   - ${check}`)
    })
    console.log('\n💡 Run individual tools to see detailed error messages:')
    console.log('   - npm run lint:check')
    console.log('   - npm run format:check')
    console.log('\n🔧 Or run fixes:')
    console.log('   - npm run lint (auto-fix ESLint issues)')
    console.log('   - npm run format (auto-format with Prettier)')
    process.exit(1)
  } else {
    console.log('✅ All linting checks passed!')
    console.log('\n🎉 Your code meets all quality standards!')
    process.exit(0)
  }
}

main()
